test/extensions/filters/listener/common/fuzz/listener_filter_fuzzer.cc

```diff
@@ -10,21 +8,92 @@ void ListenerFilterFuzzer::fuzz(
     Network::ListenerFilter& filter,
     const test::extensions::filters::listener::FilterFuzzTestCase& input) {
   try {
-    fuzzerSetup(input);
+    socket_.setLocalAddress(Network::Utility::resolveUrl(input.sock().local_address()));
   } catch (const EnvoyException& e) {
-    ENVOY_LOG_MISC(debug, "EnvoyException: {}", e.what());
-    return;
+    // Socket's local address will be nullptr by default if fuzzed local address is malformed
+    // or missing - local address field in proto is optional
+  }
+  try {
+    socket_.setRemoteAddress(Network::Utility::resolveUrl(input.sock().remote_address()));
+  } catch (const EnvoyException& e) {
+    // Socket's remote address will be nullptr by default if fuzzed remote address is malformed
+    // or missing - remote address field in proto is optional
+  }
+
+  FuzzedHeader header(input);
+
+  if (!header.empty()) {
+    ON_CALL(os_sys_calls_, recv(kFakeSocketFd, _, _, MSG_PEEK))
+        .WillByDefault(testing::Return(Api::SysCallSizeResult{static_cast<ssize_t>(0), 0}));
+
+    ON_CALL(dispatcher_,
+            createFileEvent_(_, _, Event::FileTriggerType::Edge,
+                             Event::FileReadyType::Read | Event::FileReadyType::Closed))
+        .WillByDefault(testing::DoAll(testing::SaveArg<1>(&file_event_callback_),
+                                      testing::ReturnNew<NiceMock<Event::MockFileEvent>>()));
   }
 
   filter.onAccept(cb_);
+
+  if (file_event_callback_ == nullptr) {
+    // If filter does not call createFileEvent (i.e. original_dst and original_src)
+    return;
+  }
+
+  if (!header.empty()) {
+    {
+      testing::InSequence s;
+
+      EXPECT_CALL(os_sys_calls_, recv(kFakeSocketFd, _, _, MSG_PEEK))
+          .Times(testing::AnyNumber())
+          .WillRepeatedly(Invoke(
+              [&header](os_fd_t, void* buffer, size_t length, int) -> Api::SysCallSizeResult {
+                return header.next(buffer, length);
+              }));
+    }
+
+    bool got_continue = false;
+
+    ON_CALL(cb_, continueFilterChain(true))
+        .WillByDefault(testing::InvokeWithoutArgs([&got_continue]() { got_continue = true; }));
+
+    while (!got_continue) {
+      if (header.done()) { // End of stream reached but not done
+        file_event_callback_(Event::FileReadyType::Closed);
```

Comment:
If there is no more data to fuzz, should this be an early return? (Does "not done" mean that the input data has not been read completely, even if the filter will do a no-op?)

Reply:
No, "not done" should handle the case where all of the fuzzed input data has been read (end of stream reached), but the parser fails to determine http. In this case, the file event will fallback to non-http and call `cb_->continueFilterChain(true)`, but it needs to receive `Event::FileReadyType::Closed` in order to trigger that behavior. The file event itself should be able to handle all other possible cases (socket read error, early return after success, waiting for the next event, etc.).

Reply:
Ahhhhh! Gotcha -- thanks so much for the explanation though. Sounds good for fixing later :)